Cleanup expired schedules

itirabasso opened the pull request.

itirabasso committed:
clean up expired reward schedules in fund

This should reduce the computational cost of `calculateLockedShares`
because if we remove expired schedules, which are fully unlocked at that
stage, we need to iterate less elements during the calculation.

remove length optimization

plaintextpaco commented on src/contracts/aludel/AludelV3Lib.sol:
```diff
@@ -244,4 +246,25 @@ library AludelV3Lib {
+
+            // check if the period is already expired
+            if (block.timestamp - schedule.start >= schedule.duration) {
+                // length decreces
```
https://www.learncpp.com/cpp-tutorial/comments/

these comments state that which is obvious from looking at the code, they're better suited to explain _why_ the code does what it does

plaintextpaco commented on src/contracts/aludel/AludelV3Lib.sol:
```diff
@@ -244,4 +246,25 @@ library AludelV3Lib {
+    function cleanupExpiredRewardSchedules(
+        IAludelV3.AludelData storage aludel
+    ) internal {
+        // check : measure how much gas this actually saves
```
looks like something to implement in this PR :upside_down_face:

plaintextpaco reviewed (changes requested):
this looks good but it's incomplete.

we should include tests like:

- a scenario which causes a locked contract with the previous implementation, but which can be recoverable with this one
- assert internal state of the rewardschedule array
- perhaps unitarily test the function

also, IMO the cleanup function should be callable directly, with a bound on the amount of iterations, so really bad situations can be recovered from (since adding a rewardSchedule is O(1))

itirabasso committed:
add single unstake test

itirabasso committed:
update gas-snapshot

itirabasso committed:
add some tests

* test_fund_without_approval
* test_fund_remove_expired_schedules

itirabasso committed:
update hardhat test

expired reward schedules are removed after the program is funded again.

itirabasso committed:
smol optimization

if we avoid accessing the array's length this is how the gas-snapshots
changes

```
test_many_users_multiple_stakes() (gas: -211 (-0.013%))
test_funding_shares() (gas: -592 (-0.094%))
test_single_unstake(uint8,uint40,uint256,uint256) (gas: 701603 (6.979%))
Overall gas change: 700800 (2.337%)
```

itirabasso committed:
add aludel fund tests

itirabasso committed:
forge fmt

plaintextpaco commented on src/contracts/aludel/AludelV3Lib.sol:
```diff
@@ -244,4 +244,25 @@ library AludelV3Lib {
+            IAludelV3.RewardSchedule memory schedule = aludel.rewardSchedules[index];
+
+            // we manage the length of the array manually to avoid accessing the array's length multiple times
+            length--;
```
you're reducing the length *unconditionally* yet conditionally removing elements. This'll cause the function to iterate over only the first half of the array.

Please add a testcase triggering this error (a scenario where a RewardSchedule should be removed but isn't) and then fix it.

plaintextpaco commented on src/test/AludelV3.t.sol:
```diff
@@ -296,4 +270,359 @@ contract AludelV3Test is Test {
+    }
+
+    function test_fund_without_approval(uint64 reward, uint32 duration) public {
+        vm.assume(reward > 0);
```
we have a test for a duration of zero but not for a reward of zero

itirabasso replied:
That's because we allow to fund 0 tokens.
I created an issue (#57) to track it.

plaintextpaco commented on src/test/AludelV3.t.sol:
```diff
@@ -296,4 +270,359 @@ contract AludelV3Test is Test {
+
+        IAludelV3.AludelData memory data = aludel.getAludelData();
+        // before fund there should be no schedules
+        assertEq(data.rewardSchedules.length, 0);
```
this looks like an assert that you added when playing around with the codebase to see what would work, and not one conclusive to test_fund_remove_expired_schedules

plaintextpaco commented on src/test/AludelV3.t.sol:
```diff
@@ -296,4 +270,359 @@ contract AludelV3Test is Test {
+        vm.warp(block.timestamp + 1);
+        aludel.fund(reward, duration);
+        // so there should be two schedules
+        assertEq(aludel.getAludelData().rewardSchedules.length, 2);
```
the body of the test doesn't fully relate to its description. it states that fund should remove expired schedules yet none is removed.

plaintextpaco commented on src/test/AludelV3.t.sol:
```diff
@@ -296,4 +270,359 @@ contract AludelV3Test is Test {
+    }
+}
+
+contract Given_A_New_Aludel_Is_Being_Funded is Test {
```
This contract has a lot of code duplication with AludelV3Test

Please create a 3rd contract with the setup common to both and make them extend the setUp method and add whatever extra state variables are necessary.

itirabasso replied:
I agree with this, but I was hesitant to add it right away.
Would you agree to have a few more iterations of these test contracts before creating the common contract for all tests?

plaintextpaco replied:
I'm not saying we should have a common contract for *all* tests, that feels like premature optimization given the current state of affairs, I was talking about just this testfile.

If you look at both contracts side by side, you can see the repetition easily:
![image](https://user-images.githubusercontent.com/111788132/223143510-32ec9e43-8768-4c85-bcd3-632cad4f2689.png)

I don't like the idea of merging so much repeated code, but I wouldn't mind it in a draft PR

plaintextpaco commented on src/test/AludelV3.t.sol:
```diff
@@ -296,4 +270,359 @@ contract AludelV3Test is Test {
+        aludel.fund(reward, duration);
+    }
+
+    function test_revert_when_aludel_has_not_enough_allowance(uint256 reward, uint256 duration, uint256 allowance)
```
these tests are nice to have but they aren't really related to the cleanup of reward schedules. They could've been done in a separate PR and keep things a bit tidier

itirabasso replied:
Although I understand with what you're saying, I don't see the problem really.
Actually in this case if I had to create another branch just to add certain tests it would just killed my workflow.
I would agree with you if instead of tests these changes were in the implementation, but because the cleanup is only executed when `fund` is called I would even argue that making these changes here is desirable.

I can compromise that to keep it strictly separate in the future.

plaintextpaco replied:
In the future what you could do is, when you realize you're making changes that don't fit in the current PR but are nevertheless worthwhile:

- get the changes into a separate commit
- get the commit out of the current branch
- cherrypick the commit into a separate branch later

plaintextpaco commented on src/test/AludelV3.t.sol:
```diff
@@ -296,4 +270,359 @@ contract AludelV3Test is Test {
+        aludel.fund(reward, 0);
+    }
+
+    // function test_when_fund_params_are_valid_and_fund_is_called_multiple_times_then_adds_new_schedule(uint256 reward, uint256 duration) public callerOwner {
```
whoops

plaintextpaco commented on src/test/AludelV3.t.sol:
```diff
@@ -296,4 +270,359 @@ contract AludelV3Test is Test {
+    {
+        reward = bound(reward, 1, MAX_REWARD);
+        duration = bound(duration, 1, 1000 days);
+
```
why all the whitespace?

itirabasso replied:
I personally don't mind it.
if neither forge fmt nor lint-sol cries about it, I'll just go with it.

plaintextpaco replied:
for me the whitespace in source code contributes to how I understand it, I expect empty lines to have a similar function to the paragraph separation in natural languages, and when I see two emtpy lines together I expect some big context switch, such as in sections in a book. (or, say, two different contracts, or public vs internal functions of the same contract)

Having the whitespace mean nothing at all (such as having 1-2 emtpy lines between code that accomplishes the same thing) adds a bit of cognitive noise to processing code for me.

this, for example, is how this function makes most sense to me:
```
        reward = bound(reward, 1, MAX_REWARD);
        duration = bound(duration, 1, 1000 days);
        Utils.fundMockToken(admin.addr(), rewardToken, reward);
        rewardToken.approve(address(aludel), reward);

        aludel.fund(reward, duration);
        assertEq(aludel.getAludelData().rewardSharesOutstanding, reward * BASE_SHARES_PER_WEI);
```

plaintextpaco reviewed (changes requested):
issues blocking merging are:

- modularization of tests
- lack of tests regarding schedule removal, which left bugs slip through

itirabasso commented on src/test/AludelV3.t.sol:
```diff
@@ -28,9 +28,9 @@ import {Utils} from "./Utils.sol";
 import {UserFactory} from "./UserFactory.sol";
 
 import "forge-std/console2.sol";
+import "forge-std/StdUtils.sol";
```
I understand this concern for contracts that are deployed but I don't see the point of enforcing this in tests.
I would prefer to have the whole package available instead of having to specify each of the names I want to use.

plaintextpaco replied:
you can do `import "forge-std/StdUtils" as STD` and access every exported thing as `STD.thing`

this'd not pollute the global namespace, and would make explicit where every usage comes from. In that way, if I want to find the docs for a particular function when reviewing the code or taking a look at it some time later, I can easily figure out where to look, whereas if the name is in the global namespace I'd have to do a fair bit more guessing
